stylix/darwin: add module

auscyber opened the pull request:
adds a module for wallpaper and colours for darwin



<!-- Describe your PR above, following Stylix commit conventions. -->

---

<!--
Unless otherwise specified, the following checkboxes are not mandatory, but
drastically accelerate the reviewing and merging process of this PR.
-->
- [ ] <!-- MANDATORY --> I certify that I have the right to submit this contribution under the [MIT license](https://github.com/nix-community/stylix/blob/master/LICENSE)
- [ ] Commit messages adhere to [Stylix commit conventions](https://nix-community.github.io/stylix/commit_convention.html)
- [ ] Theming changes adhere to the [Stylix style guide](https://nix-community.github.io/stylix/styling.html)
- [ ] Changes have been [tested locally](https://nix-community.github.io/stylix/modules.html#development-setup)
- [ ] Changes have been [tested in testbeds](https://nix-community.github.io/stylix/testbeds.html)
- [ ] Each commit in this PR is suitable for backport to the current stable branch

auscyber committed:
stylix/darwin: add module

adds a module for wallpaper and colours for darwin

auscyber committed:
stylix/darwin: add module

adds a module for wallpaper and colours for darwin

auscyber committed:
fixup! stylix/darwin: add module

auscyber committed:
fixup! stylix/darwin: add module

0xda157 reviewed (changes requested):
this should probably go under `/stylix/darwin/` instead of `/modules/darwin/`

> i'm getting some errors to do with callPackage not existing

this is probably due to the doc eval, cc @MattSturgeon

auscyber commented:
the entire thing? or just the desktoppr?

MattSturgeon commented on modules/darwin/hm.nix:
```diff
@@ -0,0 +1,53 @@
+mkTarget {
+  name = "darwin";
+  humanName = "Darwin";
+  extraOptions =
+    {
+      image,
+    }:
+    {
+      home.activation = {
+        stylixBackground = ''
+          run ${pkgs.callPackage ./desktoppr.nix { }}/bin/desktoppr all ${image}
+
+        '';
+      };
+    };
```
> > i'm getting some errors to do with callPackage not existing
> 
> this is probably due to the doc eval, cc @MattSturgeon

`extraOptions` is for declaring options. This looks a lot more like a config definition (you haven't used `lib.mkOption` or any of its wrappers).

The reason `pkgs.callPackage` doesn't exist when declaring options is to ensure it is not accidentally used in option docs (i.e. option examples, defaultText, etc). See #1212

auscyber replied:
i'm familiar with the distinction and have written many modules, i was just very unsure how this mkStylixModule system works

auscyber committed:
fixup! fixup! stylix/darwin: add module

auscyber committed:
fixup! fixup! stylix/darwin: add module

Eveeifyeve commented on stylix/darwin/desktoppr.nix:
```diff
@@ -0,0 +1,37 @@
+{
+  stdenv,
+  fetchurl,
+  lib,
+  unzip,
+}:
+stdenv.mkDerivation {
+  name = "desktoppr";
+  version = "0.5-218";
+  pname = "desktoppr";
+  src = fetchurl {
+    url = "https://github.com/scriptingosx/desktoppr/releases/download/v0.5/desktoppr-0.5-218.zip";
+    hash = "sha256-Oa9gAQjOaJHYyT5JBUiFCxL1sQP1dqlFBm+GdmLHNNM=";
+  };
+
+  buildInputs = [
+    unzip
+  ];
+
+  unpackPhase = ''
+    runHook preUnpack
+
+    unzip $src
+
+    runHook postUnpack
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    cp -r desktoppr $out/bin
+
+    runHook postInstall
+  '';
+
+}
```
This pr might be worth mentioning https://github.com/NixOS/nixpkgs/pull/397284

Eveeifyeve commented on modules/darwin/hm.nix:
```diff
@@ -0,0 +1,53 @@
+
+        targets.darwin.defaults."Apple Global Domain" = {
+          AppleIconAppearanceCustomTintColor = builtins.readFile "${
+            pkgs.runCommand "color" { } ''
+              full=$((16#${colors.base00}))
+              red=$((($full >> 16) & 0xff))
+              red=$(echo "scale=10; $red.0 / 255.0 " | bc |awk '{printf "%f", $0}')
+              green=$((($full >> 8) & 0xff))
+              green=$(echo "scale=10; $green.0 / 255.0  " | bc | awk '{printf "%f", $0}')
+              blue=$(($full & 0xff))
+              blue=$(echo "scale=10; $blue.0 / 255.0 " | bc | awk '{printf "%f", $0}')
+              echo "$red $green $blue 1.0" > $out
+            ''
```
Like in colors.nix module that was recently merged for `mkHexOpacityColor` and `mkHexColor`

Eveeifyeve commented on modules/darwin/hm.nix:
```diff
@@ -0,0 +1,53 @@
+        };
+      }
+    )
+  ];
```
I would like to mention there is a home-manager module [that you can use now instead](https://search.nixos.org/options?channel=25.11&include_home_manager_options=1&include_modular_service_options=1&include_nixos_options=1&query=desktoppr&show=home-manager-option:programs.desktoppr.enable)

philocalyst replied:
Ah well

Eveeifyeve commented on stylix/darwin/desktoppr.nix:
This package is no longer needed [as it's packaged in nixpkgs.](https://search.nixos.org/packages?channel=unstable&include_home_manager_options=1&include_modular_service_options=1&include_nixos_options=1&query=desktoppr&show=desktoppr)

philocalyst replied:
I swear I searched directly for this, wow, thanks for the look-over

Eveeifyeve reviewed (changes requested):
I would like to eventually as many people would, see this pr getting merged. However there is some major improvements that could be done, before I can test and approve.

trueNAHO reviewed (changes requested):
> I would like to eventually as many people would, see this pr getting merged. However there is some major improvements that could be done, before I can test and approve.

Considering recent upstream advancements, this PR could be substantially simplified. @Eveeifyeve, feel free to take over if @auscyber does not respond within a week.

---

> They didn't get back to my PR or email :(

Are you talking about a Stylix PR? You do not seem to have opened any Stylix PRs: https://github.com/nix-community/stylix/issues?q=involves%3Aphilocalyst.

philocalyst committed:
darwin/tint: add macOS accent tint module

Adds a home-manager target that sets the macOS system accent tint colour
using the base00 palette entry via targets.darwin.defaults.

philocalyst committed:
darwin/desktoppr: add macOS wallpaper module

Adds a home-manager target that sets the desktop wallpaper and scaling
mode on macOS using pkgs.desktoppr from nixpkgs. Auto-enables on Darwin.

philocalyst committed:
darwin: add module entrypoint and metadata

Wires up tint and desktoppr sub-modules via lib.modules.importApply.
Adds metadata with maintainers auscyber and philocalyst.

auscyber committed:
darwin: desktoppr fix wrong names

auscyber committed:
Merge pull request #1 from philocalyst/master

trueNAHO commented on modules/darwin/tint.nix:
```diff
@@ -0,0 +1,32 @@
+    ];
+in
+mkTarget {
+  name = "darwin";
+  humanName = "Darwin";
```
This is unnecessary after resolving https://github.com/nix-community/stylix/pull/1892#discussion_r3364189758:

```suggestion
```

trueNAHO commented on modules/darwin/tint.nix:
```diff
@@ -0,0 +1,32 @@
+
+  config =
+    { colors }:
+    lib.mkIf pkgs.stdenv.hostPlatform.isDarwin {
```
This should already be implied by `autoEnable`:

```suggestion
    {
```

auscyber replied:
unfortunately it is not, and ci fails

trueNAHO commented on modules/darwin/tint.nix:
```diff
@@ -0,0 +1,32 @@
+      mkChannel =
+        start:
+        builtins.toJSON (
+          (lib.fromHexString (builtins.substring start 2 color)) / 255.0
```
Does this require padding or does it gracefully work? See https://github.com/nix-community/stylix/pull/2310/changes/687b9af54dff298a4de86936077c7e5cb25a5622 as an example.

trueNAHO commented on modules/darwin/tint.nix:
```diff
@@ -0,0 +1,32 @@
+  config =
+    { colors }:
+    lib.mkIf pkgs.stdenv.hostPlatform.isDarwin {
+      targets.darwin.defaults."Apple Global Domain".AppleIconAppearanceCustomTintColor =
```
Consider reducing the scope of the `mkDarwinColor` local variable:

```suggestion
      targets.darwin.defaults."Apple Global Domain".AppleIconAppearanceCustomTintColor = let
        mkDarwinColor = /* ... */;
      in
```

trueNAHO commented on modules/darwin/tint.nix:
```diff
@@ -0,0 +1,32 @@
+          (lib.fromHexString (builtins.substring start 2 color)) / 255.0
+        );
+    in
+    builtins.concatStringsSep " " [
```
```suggestion
    toString [
```

trueNAHO commented on flake/dev/pre-commit.nix:
```diff
@@ -40,6 +40,7 @@
             settings.config.default.extend-identifiers = lib.genAttrs [
               "MrSom3body"
               "substituters"
+              "Wirth"
```
Keep this sorted:

```suggestion
              "Wirth"
              "substituters"
```

trueNAHO commented on modules/darwin/desktoppr.nix:
```diff
@@ -0,0 +1,30 @@
+          picture = image;
+        };
+      };
+
```
Nitpick:

```suggestion
```

trueNAHO commented on modules/darwin/desktoppr.nix:
```diff
@@ -0,0 +1,30 @@
+  config =
+    { image, imageScalingMode }:
+    {
+      programs.desktoppr = {
+        enable = pkgs.stdenv.hostPlatform.isDarwin;
+        settings = {
+          scale = mkScalingMode imageScalingMode;
+          picture = image;
+        };
+      };
```
The `enable` should already be implied by `autoEnable`:

```suggestion
      programs.desktoppr.settings = {
        scale = mkScalingMode imageScalingMode;
        picture = image;
      };
```

auscyber replied:
unfortunately no, it needs to be forcibly disabled on linux

trueNAHO commented on modules/darwin/desktoppr.nix:
```diff
@@ -0,0 +1,30 @@
+mkTarget:
+{ lib, pkgs, ... }:
+let
+  mkScalingMode =
```
Inline this local variable.

trueNAHO commented on modules/darwin/hm.nix:
```diff
@@ -0,0 +1,7 @@
+{ mkTarget, lib, ... }:
+{
+  imports = map (module: lib.modules.importApply module mkTarget) [
+    ./tint.nix
+    ./desktoppr.nix
+  ];
```
While this very cleverly creates two seperate `stylix.targets.{darwin,desktoppr}` namespaces under the `"Darwin"` module, this IMHO abuses the current freedom of our custom module system with manual `mkTarget` calls, which may become stricter in the future. Consider creating a single namespace, like all other existing Stylix modules.

trueNAHO replied:
> While this very cleverly creates two seperate `stylix.targets.{darwin,desktoppr}` namespaces under the `"Darwin"` module, this IMHO abuses the current freedom of our custom module system with manual `mkTarget` calls, which may become stricter in the future. Consider creating a single namespace, like all other existing Stylix modules.

Alternatively, what about creating seperate `/modules/{darwin,desktoppr}/hm.nix` targets?
